### PR TITLE
feat(cli): add support for multiple sources to app diff|manifests command with `revisions` flag

### PR DESCRIPTION
**Pull request by ishitasequeira**
The PR adds support for multiple sources to `argocd app manifests` and `argocd app diff` commands. As part of this PR, the support for multiple sources is by the addition of 2 new flags `--revisions` and `--source-indexes`. This 2 flags allow the user to set a revision for a specific source defined by index in `source-indexes` flag.

Examples:
1. `argocd app manifests` command
```
./dist/argocd app manifests guestbook --revisions 4171e73dec711bde9a130398d3cfc93363b16339 --source-indexes 1
```

2. `argocd app diff` command
```
./dist/argocd app diff guestbook --revisions 4171e73dec711bde9a130398d3cfc93363b16339 --source-indexes 1
```

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: cmd/argocd/commands/app.go [hunk 1]
```diff
@@ -1156,7 +1163,27 @@ func NewApplicationDiffCommand(clientOpts *argocdclient.ClientOptions) *cobra.Co
 			argoSettings, err := settingsIf.Get(ctx, &settings.SettingsQuery{})
 			errors.CheckError(err)
 			diffOption := &DifferenceOption{}
-			if revision != "" {
+			if app.Spec.HasMultipleSources() && len(revisions) > 0 && len(sourceIndexes) > 0 {
+
+				revisionSourceMappings := make(map[int64]string, 0)
+				for i, index := range sourceIndexes {
+					if index < 0 {
+						errors.CheckError(fmt.Errorf("source-index cannot be less than or equal to 0. Index starts at 1."))
```
Comment (agaudreault) [inline cmd/argocd/commands/app.go:1171]: People using ArgoCD are developers and are used to index starting at 0 IMO.

File: docs/user-guide/commands/argocd_app_diff.md [hunk 2]
```diff
@@ -27,7 +27,9 @@ argocd app diff APPNAME [flags]
       --local-repo-root string      Path to the repository root. Used together with --local allows setting the repository root (default "/")
       --refresh                     Refresh application data when retrieving
       --revision string             Compare live app to a particular revision
+      --revisions stringArray       Show manifests at specific revisions (comma-separated) for the index of sources in source-indexes
       --server-side-generate        Used with --local, this will send your manifests to the server for diffing
+      --source-indexes int64Slice   List of source indexes. Default is empty array. Indexes start at 1. (default [])
```
Comment (agaudreault) [inline docs/user-guide/commands/argocd_app_diff.md:32]: `intArray` type would make it clearer to users  and not leak go implementation details

File: go.mod [hunk 3]
```diff
@@ -94,7 +94,7 @@ require (
 	gopkg.in/yaml.v3 v3.0.1
 	k8s.io/api v0.26.11
 	k8s.io/apiextensions-apiserver v0.26.10
-	k8s.io/apimachinery v0.26.11
+	k8s.io/apimachinery v0.29.3
```
Comment (agaudreault) [inline go.mod:97]: Is that necessary in this PR? IIRC, updating this affect the kubernetes compatibility. It's probably better to update all k8s.io dependency to be in v0.29 in a dedicated PR if necessary.

**Review by agaudreault — COMMENTED** [review]
I like the input parameters, seems like a good and intuitive approach for people using multi sources

File: cmd/argocd/commands/app.go (hunk 1 above)
Comment (ishitasequeira) [reply cmd/argocd/commands/app.go:1171]: I had initially started with 0 as the index, but we updated it to 1 as the cli users all might not be developers only.

File: docs/user-guide/commands/argocd_app_diff.md (hunk 2 above)
Comment (ishitasequeira) [reply docs/user-guide/commands/argocd_app_diff.md:32]: Unfortunately, this is an auto-generated document. I am using `int64Slice` datatype for sourceIndexes as we do not have a function to parse `int[]` input in flagSet.

File: go.mod (hunk 3 above)
Comment (ishitasequeira) [reply go.mod:97]: Will revert this.

File: cmd/argocd/commands/app.go (hunk 1 above)
Comment (anandf) [reply cmd/argocd/commands/app.go:1171]: I too had a similar view. Something named Index, users may assume that it starts from 0. If we name the argument differntly like ordinal or position, may be we can emphasise that it starts from 1.

File: cmd/argocd/commands/app.go [hunk 4]
```diff
@@ -1125,6 +1125,8 @@ func NewApplicationDiffCommand(clientOpts *argocdclient.ClientOptions) *cobra.Co
 		serverSideGenerate bool
 		localIncludes      []string
 		appNamespace       string
+		revisions          []string
+		sourceIndexes      []int64
```
Comment (anandf) [inline cmd/argocd/commands/app.go:1129]: Should this be `int64` ? I feel we can use `uint` or `unit32` here as the index positions do not have negative numbers.

Comment (ishitasequeira) [reply cmd/argocd/commands/app.go:1129]: The command does not support `uint` or `uint32`. Thus, had to keep it `int` or `int64`. Had used `int64` to keep it consistent with what is needed by `revisionSourceMappings`. I can change this to `int` and then typecast it to `int64` for creating `revisionSourceMappings`.

Comment (anandf) [reply cmd/argocd/commands/app.go:1129]: If flagset support is missing for `uint`, then let's stick to `int64`

**Review by anandf — APPROVED** [review]
LGTM

File: cmd/argocd/commands/app.go (hunk 1 above)
Comment (agaudreault) [reply cmd/argocd/commands/app.go:1171]: Yeah I think consistency is more important for this PR, but I think this should be quickly reassessed before this it is used and released.

**Review by agaudreault — APPROVED** [review]
LGTM. Lets discuss the index at the next [contributor meeting](https://github.com/argoproj/argo-cd/blob/affd1cb2517d3fedd957edf11dba73b9cb9dfe5a/docs/developer-guide/code-contributions.md?plain=1#L102)

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Can you add a `Long` usage description for the `app manifests` command to provide an example of how to use this?

I'm not sure the docs are correct about the flags being comma-delimited:

```
It's behaving like it's not comma-delimited but rather repeated:
dist/argocd app manifests test --revisions d7927a27b4533926b7d86b5f249cd9ebe7625e90,53e28ff20cc530b9ada2173fbbd64d48338583ba --source-indexes 1,2
FATA[0000] While using revisions and source-indexes, length of values for both flags should be same.
```

This seems to work:

```
dist/argocd app manifests test --revisions d7927a27b4533926b7d86b5f249cd9ebe7625e90 --revisions 53e28ff20cc530b9ada2173fbbd64d48338583ba --source-indexes 1 --source-indexes 2
```

Should we add an e2e test to make sure the usage is tested?

**Review by crenshaw-dev — APPROVED** [review]
LGTM!

**Pull request merged**